Product Collection: Inspector control to change selected product

imanish003 opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #44877.

This is a follow up PR to #50164. In this PR, I have implemented Inspector control to change selected product. This way, merchant can change the selected product using Inspector control. Here is how it looks like:
![image](https://github.com/user-attachments/assets/aeb4e762-5c22-4e9f-8df9-053793a3c93f)

Once user click on selected product button, it will show a Popover to search and select a different product. As soon as a new product will be selected, it will reflect in the inspector control:
![image](https://github.com/user-attachments/assets/d326cfc8-589a-45f0-81fb-4612bc562fb9)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

**Pre-requisite:** Install this plugin - 
[register-product-collection-tester.zip](https://github.com/user-attachments/files/16362323/register-product-collection-tester.zip)

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Case 1: Adding "My Custom Collection - Product Context" to a place where product context doesn't exist**

1. Create a new post and add Product Collection block.
2. Choose "My Custom Collection - Product Context" block.
3. Choose any product from Editor product picker.
4. Verify that the selected product is shown in Inspector control.
![image](https://github.com/user-attachments/assets/d6e75174-6f8d-4549-9438-ac58bc531680)
5. Click on selected product and verify that a Popover is shown to search and select a different product.
6. Search and select a different product and verify that it reflects in Inspector control.
7. _To be verified by developers only_: 
  - In the code editor, verify that `productReference` attribute is updated with new product id of selected product. When a product is changed, `productReference` attribute should be updated with new product id.

**Case 2: Adding "My Custom Collection - Product Context" on Single Product template**

1. Go to Single Product template and add Product Collection block.
2. Choose "My Custom Collection - Product Context" block.
3. Verify that product picker is not shown in Inspector control.

**Case 3: Adding "My Custom Collection - Product Context" on Specific Product template**

1. Create a new Specific Product template (See screenshot below)
![image](https://github.com/user-attachments/assets/7cbfe19e-e97e-4ca7-988c-32a8ee1a3fd4)

2. Go to specific product template that you created in step 1 and add Product Collection block.
3. Verify that product picker is not shown in Inspector control.

**Case 4: Collection which doesn't require product context, shouldn't show product picker**

1. Create a new post and add Product Collection block.
2. Choose any other collection except "My Custom Collection - Product Context".
3. Verify that product picker is not shown in Inspector control.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection - Implement Inspector control to change selected product
</details>

When a collection require a product context, but if it doesn't exist in current template/page/post etc. then we will show a control in inspector controls to change the selected product. This way, merchant can manually provide a product context to the collection.

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Add Linked Product Control to Product Collection Block Inspector Controls

- Introduced a new `LinkedProductControl` component in the Product Collection block's Inspector Controls.
- This control allows users to link a specific product to the product collection via a dropdown with a search capability.
- Added corresponding styles to `editor.scss`.
- Integrated a `useGetProduct` hook in the `utils.tsx` to fetch and manage the state of the linked product data, including handling loading states and errors.
- Updated the Inspector Controls to include the new Linked Product Control component, enhancing the block's customization options for users.

imanish003 committed:
Add E2E tests

imanish003 committed:
Hide product picker when product context is available

imanish003 committed:
Improve logic to hide Linked Product Control

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,138 @@
+	usesReference: string[] | undefined;
+} ) => {
+	const [ isDropdownOpen, setIsDropdownOpen ] = useState< boolean >( false );
+	const { product, isLoading } = useGetProduct( query.productReference );
```
What about `hasError`? I think we should either consume it or simply remove it from `useGetProduct` if there's no use for it yet.

imanish003 replied:
Good catch! I thought I will use it but didn't, so I removed it 🤝

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,138 @@
+	const [ isDropdownOpen, setIsDropdownOpen ] = useState< boolean >( false );
+	const { product, isLoading } = useGetProduct( query.productReference );
+
+	const isShowLinkedProductControl = useMemo( () => {
```
```suggestion
	const showLinkedProductControl = useMemo( () => {
```

I think it would be easier to read without `is`

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,138 @@
+		);
+	}, [ location.type, query?.productReference, usesReference ] );
+
+	const jsxProductButton = ( {
+		isOpen,
+		onToggle,
+	}: {
+		isOpen: boolean;
+		onToggle: () => void;
+	} ) => {
+		if ( isLoading && ! product ) {
+			return <Spinner />;
+		}
+
+		return (
+			<Button
+				className="wc-block-product-collection-linked-product-control__button"
+				onClick={ onToggle }
+				aria-expanded={ isOpen }
+				disabled={ isLoading }
+			>
+				<Flex direction="row" expanded justify="flex-start">
+					<FlexItem className="wc-block-product-collection-linked-product-control__image-container">
+						<img
+							src={ product?.images?.[ 0 ]?.src }
+							alt={ product?.name }
+						/>
+					</FlexItem>
+					<Flex direction="column" align="flex-start" gap={ 0 }>
+						<FlexItem>
+							<Text color="inherit">{ product?.name }</Text>
+						</FlexItem>
+						<FlexItem>
+							<Text color="inherit">{ product?.sku }</Text>
+						</FlexItem>
+					</Flex>
+				</Flex>
+			</Button>
+		);
+	};
```
What do you think about extracting it outside of this function to the top? It's a regular React component so it could be `const ProductButton = ...`

imanish003 replied:
Makes sense, I will extract it to the top.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,138 @@
+		);
+	};
+
+	const jsxPopoverContent = (
+		<ProductControl
+			selected={ query?.productReference as SelectedOption }
+			onChange={ ( value = [] ) => {
+				const isValidId = ( value[ 0 ]?.id ?? null ) !== null;
+				if ( isValidId ) {
+					setAttributes( {
+						query: {
+							...query,
+							productReference: value[ 0 ].id,
+						},
+					} );
+					setIsDropdownOpen( false );
+				}
+			} }
+			messages={ {
+				search: __( 'Select a product', 'woocommerce' ),
+			} }
+		/>
+	);
```
The same as above, I'd extract it to `PopoverContent` component outside of `LinkedProductControl` component

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,138 @@
+					className="wc-block-product-collection-linked-product-control"
+					contentClassName="wc-block-product-collection-linked-product__popover-content"
+					popoverProps={ { placement: 'left-start' } }
+					renderToggle={ ( { isOpen, onToggle } ) =>
+						jsxProductButton( { isOpen, onToggle } )
+					}
```
Can be simplified to:

```suggestion
					renderToggle={ jsxProductButton }
```

or if you extract it as a component:

```suggestion
					renderToggle={ ProductButton }
```

imanish003 replied:
I don't think it's valid because both ProductButton and LinkedProductPopoverContent accept more props, which we need to pass explicitly. Please let me know if I am missing something 🤔

kmanijak reviewed (changes requested):
I'm impressed by a recent work of yours @imanish003, awesome!

Overall, the solution works really good. I really like the flow of changing the linked product! I left some comments but nothing major.

### When name / SKU is longer than one line it's centered.

Here are examples. In second one you can see the product name is left-aligned and SKU is centered.
<img width="276" alt="image" src="https://github.com/user-attachments/assets/36cbc222-8d91-4f69-b572-c19f00907695">
<img width="269" alt="image" src="https://github.com/user-attachments/assets/fc74b33a-959e-4436-b2cc-4fc45f19425a">

### Testing steps updates

You use `selectedReference` name in testing steps instead of `productReference`. Please update it.

imanish003 committed:
Merge branch 'add/44877-context-linking-a-product-with-collection' into add/44877-context-linking-a-product-with-collection-inspector-control

imanish003 committed:
Remove hasError state from useGetProduct hook

imanish003 committed:
Rename isShowLinkedProductControl to showLinkedProductControl

imanish003 committed:
Convert jsxProductButton to ProductButton component

imanish003 committed:
Refactor jsxPopoverContent to LinkedProductPopoverContent component

imanish003 committed:
Improve UI of Linked Product Control

imanish003 commented:
Hey @kmanijak,

I have addressed all the feedback. Can you please take a look when you get a chance? 🙇‍♂️ 

Regarding the long product name and SKU, I have made some changes. Here is how it looks like now:
![image](https://github.com/user-attachments/assets/1f697107-e430-4b43-b056-2ff15ee61a0f)


It can probably be improved further, but I think it's good enough for now. Later, we can consider whether to truncate it and show the full name/SKU on hover in the tooltip. What do you think?

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,166 @@
+import ProductControl from '@woocommerce/editor-components/product-control';
+import { SelectedOption } from '@woocommerce/block-hocs';
+import { useState, useMemo } from '@wordpress/element';
+import { WooCommerceBlockLocation } from '@woocommerce/blocks/product-template/utils';
```
```suggestion
import type { WooCommerceBlockLocation } from '@woocommerce/blocks/product-template/utils';
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx:
```diff
@@ -6,8 +6,10 @@ import { addFilter } from '@wordpress/hooks';
-import { useLayoutEffect } from '@wordpress/element';
+import { useEffect, useLayoutEffect, useState } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
+import { ProductResponseItem } from '@woocommerce/types';
```
```suggestion
import type { ProductResponseItem } from '@woocommerce/types';
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,166 @@
+import { SelectedOption } from '@woocommerce/block-hocs';
+import { useState, useMemo } from '@wordpress/element';
+import { WooCommerceBlockLocation } from '@woocommerce/blocks/product-template/utils';
+import { ProductResponseItem } from '@woocommerce/types';
```
```suggestion
import type { ProductResponseItem } from '@woocommerce/types';
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -0,0 +1,166 @@
+ * Internal dependencies
+ */
+import { useGetProduct } from '../../utils';
+import {
```
```suggestion
import type {
```

kmanijak commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts:
```diff
@@ -1074,6 +1074,59 @@ test.describe( 'Editor product picker', () => {
+				admin,
+				page,
+			} ) => {
+				await admin.createNewPost();
+				await pageObject.insertProductCollection();
+				await pageObject.chooseCollectionInPost( key as Collections );
```
You can use a helper instead:

```suggestion
				await pageObject.createNewPostAndInsertBlock( key )
```

imanish003 replied:
I didn't use `createNewPostAndInsertBlock` because I added a `chooseProductInEditorProductPickerIfAvailable` call in this function, which chooses a product in the product picker. In this specific test, we don't want that to run 🙂

kmanijak reviewed (approved):
Left some minor comments. But overall, amazing work! Since it's a PR to another feature branch, I'm preapproving although tests are failing (BTW, I suggested a change in tests that may help!)

imanish003 committed:
Merge branch 'add/44877-context-linking-a-product-with-collection' into add/44877-context-linking-a-product-with-collection-inspector-control

imanish003 committed:
Address PR feedback

imanish003 committed:
Fix E2E tests
